### ISSUE TITLE
Kernel.method calls `inspect` on obj instead of class

### ISSUE THREAD

**Issue by e2** [rubinius/rubinius#3373]
It's a minor thing, but I thought it's worth reporting.

(I don't know what the spec is for Ruby on this.)

```ruby
class Foo
  def inspect
    fail NotImplementedError
  end
end

Foo.new.method(:bar)
```

#### Expected (on MRI):
```
foo.rb:7:in `method': undefined method `bar' for class `Foo' (NameError)
	from foo.rb:7:in `<main>'
```

#### Actual (on rbx-head and rbx-2.5.2 - probably on all versions)
```
An exception occurred running foo.rb:
NotImplementedError (NotImplementedError)
Backtrace:
                       Foo#inspect at foo.rb:3
                Kernel(Foo)#method at kernel/common/kernel.rb:526
                 Object#__script__ at foo.rb:7
  Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
  Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
           Rubinius::Loader#script at kernel/loader.rb:655
             Rubinius::Loader#main at kernel/loader.rb:809

```

**Comment by YorickPeterse** [comment]
This is because we use `#inspect` in the error output:

```
$ ruby /tmp/test.rb 
An exception occurred running /tmp/test.rb:

undefined method `bar' for #<Foo:0x150> (NameError)

Backtrace:

                Kernel(Foo)#method at kernel/common/kernel.rb:527
                 Object#__script__ at /tmp/test.rb:5
  Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
  Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
           Rubinius::Loader#script at kernel/loader.rb:656
             Rubinius::Loader#main at kernel/loader.rb:810
~ $ cat /tmp/test.rb 
class Foo

end

Foo.new.method(:bar)
```

It probably makes more sense for this to produce an error similar to the ones we use for undefined method calls such as `NoMethodError: undefined method `bar' on an instance of Foo.`.

**Comment by digitalextremist** [comment]
@e2 you beat me to this. Although it seems irregular that `#inspect` would be invoked the way it is, it seems our test causes this issue, so we ought to probably check for `rbx`and bypass this test for now.

@YorickPeterse do you know why this behavior changed, or when it changed?

**Comment by YorickPeterse** [comment]
@digitalextremist As far as I can tell (based on `git log -G 'def method\('`) nothing has changed, this has been around for several years already.

**Comment by e2** [comment]
I wanted to create a PR for this, but I probably won't have the time in the nearest future.

**Comment by e2** [comment]
@digitalextremist - relying on undefined behavior is the root problem here.

**Comment by digitalextremist** [comment]
@e2 agreed, except seeing whether *unknown* failures happen while a several year-old issue is resolved seems sensible. We patch around `jRuby` versions all the time. Thanks @YorickPeterse

**Comment by YorickPeterse** [comment]
@digitalextremist This particular problem only occurs when one overwrites `#inspect` to something that is not compatible with the default implementation. Is this something Celluloid is doing?

**Comment by digitalextremist** [comment]
Absolutely, in a test spec. How can we avoid unsettling the `inspect` that `rbx` is expecting? Does *any* overwrite of that method cause this? As @e2 showed, it's this:

```ruby
class Foo
  def inspect
    fail NotImplementedError
  end
end
```

**Comment by YorickPeterse** [comment]
@digitalextremist By not overwriting `Object#inspect` in a way that doesn't alter it's behaviour from the outside. That is, as long as it returns some kind of String (and not raise an error or something like that) you should be fine.

**Comment by digitalextremist** [comment]
@YorickPeterse, hmm. One thing we need to test is if a failure happens in an overriden `#inspect` -- is there any environment variable we can set, or something like that to bypass the `inspect` call behavior; if not, can that be easily added?

**Comment by YorickPeterse** [comment]
@digitalextremist What kind of use case are you trying to test by making `#inspect` raise an error? That is, what lead to the need for that behaviour and it being covered by the tests?

**Comment by YorickPeterse** [comment]
To further clarify, this works in MRI because:

1. It doesn't rely on `Object#inspect` when printing errors
2. It often does most of the work in C by using the underlying functions, completely bypassing the ones bound to Ruby

If MRI were to ever use `#inspect` directly these tests would probably fail in a similar manner. The same applies to JRuby by the way.

**Comment by digitalextremist** [comment]
Well, one right off the top of my head is that any method invocation on an object which is an actor, and that actor is dead, calling `inspect` on it will crash. We would prefer the behavior of `mri` and `jruby`.

But truly, this is not *that* important, since we are triggering a failure; it just blacks out one test for us under `rbx` ... and we could do more than bypass this, but actually anticipate `rbx` behavior itself.

**Comment by YorickPeterse** [comment]
Again, this _might_ work now but it could break in the future both on MRI and JRuby should they ever choose to use `Object#inspect` in a similar fashion as Rubinius does at the moment. Can you not simply test another, non core method instead of using `#inspect`?

**Comment by e2** [comment]
@YorickPeterse - to explain briefly what's going on (without involving you unnecessarily), Celluloid has `check()` method - to check if an object's method exists and if the arity matches. 

This is to prevent method/arity mismatches from resulting in NoMethodError and ArgumentError in the thread with the object. (The idea is to re-raise the exception in the calling thread - "as if" the exception happened there).

Philosophically, it's like making the following invocation:

```ruby
Thread.new { obj.call_non_existing_method } 
```

... fail in the main thread instead.

I'm strongly opposed to having a check like that (because of method_missing and because it forces people to implement questionable exception handling in both the calling thread - and the one with the object).

The objects (with failing `#inspect`) are user objects - not Celluloid internal classes, so this is about trying to protect the user from e.g. calling a non-existent method.

In short, users can break the convention (of having `#inspect` do more than just dump instance variables), but I see no reason why Celluoid should have to bend over backwards to protect users from this.

Other than that I agree that we shouldn't assume that MRI and JRuby won't use `Object#inspect` in the future. Consistency is nice, but with no spec there is no "correct" implementation here to begin with.

**Comment by YorickPeterse** [comment]
> In short, users can break the convention (of having #inspect do more than just dump instance variables), but I see no reason why Celluoid should have to bend over backwards to protect users from this.

It doesn't have to, and it shouldn't either. If a user overwrites `#inspect` or whatever core method in a way that breaks compatibility with its default behaviour then this is a problem of said user. As such I don't see why Celluloid would need to test for calling `#inspect`, instead it could check for any method (e.g. `some_other_method_not_existing_on_the_test_object`).

We've had similar problems in the past with certain libraries patching `Object#is_a?`, messing with `Array` internals and similar crimes against Ruby. In a few cases we can use mirrors to hide internal state (e.g. https://github.com/rubinius/rubinius/blob/master/kernel/bootstrap/array_mirror.rb) but we can't use this for every possible method out there.

**Comment by e2** [comment]
@YorickPeterse - in short, Celluloid wants to distinguish between "invocation errors" and other exceptions. 

So, Celluloid relies on `#method()` failing only with NoMethodError (calling `#inspect` breaks that "guarantee").

But if `#method()` calls `#inspect`, it can potentially fail with any error (including network errors, since these calls are also intented to be distributed - e.g. the `#inspect` could be on another machine where the actual object is. So `#inspect` may fail no just because of a user overriding it, but simply because the thread/node crashed on the other end).

> We've had similar problems in the past with certain libraries patching Object#is_a?

Celluloid has it's on implementations for a lot of the usual metaprogramming: `#respond_to?`, `#method`, `Method` (custom class), `#__send__`, `#class`, `#method_missing`, `#to_s`, `#inspect` etc.

Currently, it's tough to get `#method()` to not call `#inspect`. But even if this was fixed, it would introduce problems, because now also the Rubinius version would have to be taken into account.

I believe that expecting metaprogramming internals to behave consistently among interpreters is placing extra handcuffs on those interpreters. And to me `#inspect` is part of the metaprogramming internals (because it's relied on in many places) - so using it in exceptions seems justified to me.

However, if `#inspect` does generate a longer dump of an object, the exception message may be hard to read, but `#method` is probably very rarely called, so this won't affect many other users either way.

I reported this because of the inconsistency with other implementations, not as a bug - so feel free to close this if you think it's not worth changing.

Thanks for your time.

**Comment by YorickPeterse** [comment]
> So, Celluloid relies on #method() failing only with NoMethodError (calling #inspect breaks that "guarantee").

> But if #method() calls #inspect, it can potentially fail with any error (including network errors, since these calls are also intented to be distributed - e.g. the #inspect could be on another machine where the actual object is. So #inspect may fail no just because of a user overriding it, but simply because the thread/node crashed on the other end).

If any method call `#method` is called on could result in network activity I doubt this would work out very well. For example, our `#method` implementation also depends on `#respond_to_missing?` and possibly others. Because we implement Ruby using (mostly) Ruby there's not much we can do about this other than _not_ using Ruby.

> Celluloid has it's on implementations for a lot of the usual metaprogramming: #respond_to?, #method, Method (custom class), #__send__, #class, #method_missing, #to_s, #inspect etc.

This should be fine as long as their behaviour is compatible with the default behaviour (e.g. `#class` still returning a class and not something else). If not then there are a (close to) infinite cases where this could break depending on how exactly Rubinius would implement something.

> I reported this because of the inconsistency with other implementations, not as a bug - so feel free to close this if you think it's not worth changing.

The way I see it, there are two things going on:

1. Celluloid overwrites certain core Ruby methods to do something different from what they usually do (correct me if I'm wrong)
2. Rubinius in certain places (e.g. `#method`) expects certain methods to be their vanilla versions (e.g. `#inspect` just returning a String)

Case 1 is something we can't do anything about. At most we can move internal instance variables/helper methods into a mirror (the internal state of `Array` being an example).

Case 2 is also something we can only do little about. One of the most common patterns we currently use that's related to this is the following:

    if nil.equal?(some_argument)

    end

Here we explicitly use `NilClass#equal?` (= `BasicObject#equal?`) so that in the event of `some_argument` defining it's own `#equal?` method we still get the behaviour we'd expect. This however can only go so far.

Thinking of all this I don't think there's much we (as in, Rubinius) can do. We _can_ remove the dependency on `#inspect` from `#method` but it would have to be replaced with `#class` (to get the class name) which could in theory introduce similar problems.

**Comment by e2** [comment]
> Because we implement Ruby using (mostly) Ruby there's not much we can do about this other than not using Ruby.

Good point - you made me realize that with Rubinius users can "hook into" a lot more than in MRI or JRuby. E.g. I've had issues stubbing ENV before in JRuby (ENV is not even a object in MRI - it's a "hash-like-thing").

> This should be fine as long as their behaviour is compatible with the default behaviour

Gosh, I now realized how incomplete coverage is in such use cases. To create a bullet-proof implementation of something using e.g. `#method()`, you pretty much need to run such code against the tests in a portion of rubyspec. Otherwise, users will get surprising behavior.

E.g. Most people implementing respond_to? won't handle the second parameter.

By redefining core methods like that, you're basically trying to write your own interpreter - food for thought when considering coverage and robustness.

> Celluloid overwrites certain core Ruby methods to do something different from what they usually do (correct me if I'm wrong)

There are more or less ok for most use cases. I wouldn't say "different", but maybe "incomplete" due to being adapted for the purpose. Celluloid basically tries to implement proxies in a way where the user won't see they are using a proxy.

E.g.:

```ruby
class MyActor
  include Celluloid
end

MyActor.new.respond_to?(:foo)
```

This instantiates the class in a new thread, and then calls `:respond_to?` on the proxy, which calls the `:__send__` method with `:respond_to?` as the parameters - and that gets passed to a queue, which the new thread is monitoring. Then, through some extra logic, the `:method(:respond_to?)` is called, and checked, then the actual `:respond_to?` is called.

So Celluloid is trying to abstract away the internal marshalling and context switching - and trying to avoid exceptions being raised in the other thread context.

> but it would have to be replaced with #class (to get the class name) which could in theory introduce similar problems.

I think you nailed it. It would be solving the issue "statistically", meaning people are more likely to define their own `:respond_to?` than `:method`, and they are more likely to define their own `:inspect` than `:class`.

Celluloid does redefine `:__class__` - to hide the proxies. `:class` is called on the wrapped_object, not on the wrapping proxy. There are also one-to-many proxies, so `:class` is made to represent the final object interacted with (not the containing-proxy class).

You could say Celluloid is trying to be a "middleman" between a user calling a method - and the actual method being called.

As for using MRI as a "reference", there's also this: http://rubini.us/2014/12/31/matz-s-ruby-developers-don-t-use-rubyspec

I think it might be even worth "borrowing" some mspec tests and incorporating them into the Celluloid tests. 

The idea in Celluloid is to just add the `include Celluloid` and the class automatically becomes thread-safe and lockless (because of context switching to the same thread). Other than that, Celluloid classes should act like standard Ruby objects (to the extent possible, given the marshalling/network/threading, etc.) to avoid surprising users.

Thanks so much for the examples - realizing Celluloid is trying to be another "Ruby interpreter" really puts things into perspective.

**Comment by YorickPeterse** [comment]
No problem! Since there's not much we change here I'll be closing this issue, feel free to re-open (or pop in Gitter) if needed.

**Comment by brixen** [comment]
It may be worth noting that since JRuby+Truffle is importing pretty much all of Rubinius core libraries, you'd most likely hit the same issue at some point in the future there (unless they stop calling Ruby methods to implement Ruby).

Even though it appears generally understood in this thread, I cannot emphasize it enough that MRI has no concrete specification. They change implementation at any time they feel like it. A Ruby method that is called today could be a C function tomorrow, or vice versa.

The only sane thing to do is respect the existing contract of Ruby core methods when overriding them. Rubinius may use those methods in places that you don't realize, but so could _any_ code, both code inside your own project or code in a library you use. You may think you'll catch that with tests and coverage. You'll be wrong. Override Ruby core methods at your own peril.

**Comment by e2** [comment]
> JRuby+Truffle is importing pretty much all of Rubinius core libraries

That makes lots of sense, now that you mention it.

> I cannot emphasize it enough that MRI has no concrete specification.

Much appreciated, @brixen - it's too easy to "assume" MRI *is* a specification. Many people do that, then assuming that JRuby or Rubinius are "broken".

> You may think you'll catch that with tests and coverage.

Sad truth. Not only that, but any isolated testing needs a deeper level of metaprogramming - which means insane work and lots of new issues to deal with - just keeping the tests working.

**Comment by headius** [comment]
JRuby+Truffle is no more exempt from compatibility with MRI than JRuby. If this is a problem in the JRuby+Truffle code we would fix it.

**Comment by headius** [comment]
Responding to other JRuby mentions:

> E.g. I've had issues stubbing ENV before in JRuby (ENV is not even a object in MRI - it's a "hash-like-thing").

ENV is special mostly because it is just a hash-like shim over the top of a native construct, the process's environment variable table. Most of the behavior you see is due to Ruby trying to make nice of setenv/getenv behavioral quirks across platforms. JRuby implements ENV using an actual hash with some modifications to behave more like setenv/getenv...but those modifications may make it more difficult to monkey-patch or stub ENV.

**Comment by e2** [comment]
> Most of the behavior you see is due to Ruby trying to make nice of setenv/getenv behavioral quirks across platforms. 

I didn't know about the quirks being a factor. (In fact, after browsing the JRuby implementation, ENV seemed so "special" and unwieldy for specs, I just encapsulated it in a gem: https://github.com/e2/nenv)

@headius - just curious: do you see calling `#inspect` when `#method()` fails as an inconsistency worth fixing in Rubinius or a justifiable engine-specific "difference"? I'm guessing you'd prefer to avoid the inconsistency with MRI. (I'm asking in case I ever stumble onto related inconsistencies in JRuby).

**Comment by brixen** [comment]
Apparently there is some misunderstanding or incorrect assumptions about my comment. For clarification: https://gist.github.com/anonymous/e8c2fba84981662d2779

**Comment by headius** [comment]
@e2 Well, since you asked, here's my opinion: Yes, I see calling `#inspect` when `#method` fails as an inconsistency worth fixing, since it prevents the correct operation of `#method` on any class that defines its own incompatible `#inspect`. That seems like a pretty big restriction for a Ruby impl to make.

There are other good reasons for not calling inspect here:

* If `#inspect` has side effects (lazy, remote data, large data, sensitive data, etc) Rubinius will trigger them when raising this error from `#method`. I have filed issues against MRI's NameError for the same reason (it retains a reference to the original object and lazily inspects it...hilarious when that object is an entire ActiveRecord graph).
* If the class undefines inspect (as in blank-slate classes that retain reflective capabilities like `#method`), it will fail to produce the right errors.

Ultimately this is a bit academic because we're talking about exceptional behavior, but if you're depending on the exception to behave right (and I won't claim there aren't valid reasons) then the difference here is a problem.

**Comment by YorickPeterse** [comment]
It's worth mentioning that as I said we _can_ remove the dependency on `#inspect` (and I'll do so in a moment to get it over with). It's just that we _can't_ guard against every case where we _might_ depend on another Ruby method that _might_ be overwritten in a way that does not adhere to the original implementation's contract. Examples are somebody monkey-patching `Array#map` to always return `"bacon"`, somebody overwriting `respond_to_missing?` to simply always return `true` and so on.

Finally, and I realize this may make me sound like an ass, I'd like to ask everybody to direct discussions to either Gitter [1] or a separate issue targeting a specific part of this discussion (instead of dumping everything into a single one), lest we not clutter this particular ticket any more (also keeps the Github notification amount down a bit).

[1]: This way we don't spam up to 134 watchers on Github when < 5 of them would actually be interested.

**Comment by headius** [comment]
> It's just that we can't guard against every case where we might depend on another Ruby method that might be overwritten in a way that does not adhere to the original implementation's contract. Examples are somebody monkey-patching Array#map to always return "bacon", somebody overwriting respond_to_missing? to simply always return true and so on.

The current issue is about someone defining `#inspect` on their own class in their own way, not monkey-patching a core method to do something obviously broken. I understand wanting people to leave core alone, but in this case it seems that you're dictating how people implement their own classes.

Glad to hear you're fixing it, in any case! :-)

I'm done discussing as well...I just came to clarify the JRuby+Truffle statement and offered my opinion when asked.

**Comment by ioquatix** [comment]
Part of the issue is that we couldn't call Actor#inspect on dead actors, but this issue was fixed, so now it is fine to call #inspect even on dead actors. I believe that using object.inspect in the exception backtrace is not a bad approach. It should be fine to call inspect but I see there is a lot of contention around this issue.
